k2/csrc/array.h

```diff
@@ -1,69 +1,113 @@
 // k2/csrc/array.h
 
 // Copyright (c)  2020  Xiaomi Corporation (author: Daniel Povey)
+//                      Fangjun Kuang (csukuangfj@gmail.com)
 
 // See ../../LICENSE for clarification regarding multiple authors
 
 #ifndef K2_CSRC_ARRAY_H_
 #define K2_CSRC_ARRAY_H_
 
-#include <functional>
-#include <limits>
-#include <memory>
+#include <type_traits>
 #include <vector>
 
+#include "glog/logging.h"
+
 namespace k2 {
 
 /*
-   We will use e.g. StridedPtr<int32_t, T> when the stride is not 1, and
+   We will use e.g. StridedPtr<T*, int32_t> when the stride is not 1, and
    otherwise just T* (which presumably be faster).
 */
-template <typename I, typename T>
+template <typename Ptr, typename I = int32_t>
 struct StridedPtr {
-  T *data;
-  I stride;
-  T &operator[](I i) { return data[i]; }
+  using T = typename std::remove_pointer<Ptr>::type;
+
+  T *data;   // it is NOT owned here
+  I stride;  // in number of elements, NOT number of bytes.
+
+  T &operator[](I i) { return data[i * stride]; }
+  const T &operator[](I i) const { return data[i * stride]; }
+
   StridedPtr(T *data, I stride) : data(data), stride(stride) {}
 };
 
-/* MIGHT NOT NEED THIS */
-template <typename I, typename Ptr>
-struct Array1 {
-  // Irregular two dimensional array of something, like vector<vector<X> >
+template <typename Ptr, typename I = int32_t>
+class Array1 {
+ public:
+  // Irregular two dimensional array of something, like vector<X>
   // where Ptr is, or behaves like, X*.
   using IndexT = I;
   using PtrT = Ptr;
+  using ValueType = typename std::remove_pointer<PtrT>::type;
+
+  Array1() : begin_(0), end_(0), data_(nullptr) {}
+
+  Array1(IndexT begin, IndexT end, PtrT data)
+      : begin_(begin), end_(end), data_(data) {}
+
+  void Init(IndexT begin, IndexT end, PtrT data) {
+    begin_ = begin;
+    end_ = end;
+    data_ = data;
+  }
+
+  Ptr begin() { return data_ + begin_; }
```

Comment:
It we want to support `StridedPtr` here, actually we need to implement somethings like `iterator` for StridedPtr

Reply:
StridedPtr *is* an iterator.. just need operator + and operator ++